Add notation document where we note our conventions

nilsdeppe opened the pull request:
## Proposed changes

- With the notation mostly figured out @kidder and I decided we should start transitioning all the documentation to use the notation. Here we have a document and place in the documentation to keep track of notation and conventions.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add notation document where we note our conventions

fmahebert commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+- A DG scheme with an \f$N^{\mathrm{th}}\f$ order basis is referred to as a
+  \f$P_N\f$ scheme and converges at order \f$\mathcal{O}\left(\Delta
+  x^{N+1}\right)\f$ where \f$\Delta x\f$ is the 1d size of the element.
+- The unit normal vector to the element is denoted by \f$n_i\f$.
```
More precisely, it's the normal to the surface/boundary/face of the element. Perhaps say so, using your preferred words?

fmahebert commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+  x^{N+1}\right)\f$ where \f$\Delta x\f$ is the 1d size of the element.
+- The unit normal vector to the element is denoted by \f$n_i\f$.
+- The numerical flux boundary correction term which includes the normal vectors
+  is denoted by \f$G(F^{i,+},n_i^+,u^+,F^{i,-},n_i^-,u^-)\f$ where the \f$+\f$
```
I dislike this because everyone else denotes the numerical flux F*. Is the goal to be more distinct from the *F dual of Faraday?

nilsdeppe replied:
No, this is to be correct. Using F* is incorrect when you have a dynamical metric (and I think with hp-adaptivity as well). It's also not true that everyone else uses F*. All the ADER papers use G, so somewhere around 60% of (G)RMHD literature actually uses G.

fmahebert commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+      p^* = & p + \frac{b^2}{2}=p+p_{\mathrm{mag}}.
+  \f}
+  We denote the conserved variables multiplied by \f$\sqrt{\gamma}\f$ with a
+  tilde on top. For example, \f$\tilde{D}=\sqrt{\gamma}D\f$.
```
How about adding a section with our notational conventions for Strahlkorpers? As I recall, those conventions come up in code review every now and then, so writing them here could be very useful.

nilsdeppe replied:
Yes, but I was hoping @markscheel could do this since he's most familiar with them. I'm trying to get what is in the current paper draft into the code.

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+  \f{align*}{
+      \mathbf{J}=\frac{\partial \mathbf{x}}{\mathbf{\xi}}
+  \f}
+  where \f$\mathbf{x}\f$ are the coordinates mapped to by coordinate map.
```
by the?

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+  \f$\mathbf{\xi}=\{\xi,\eta,\zeta\}\f$.
+- The Jacobian matrix is defined as
+  \f{align*}{
+      \mathbf{J}=\frac{\partial \mathbf{x}}{\mathbf{\xi}}
```
Use indices to make clear their order?

nilsdeppe replied:
This is what is being done in the paper.

nilsdeppe replied:
I also don't see how this is at all not clear.

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+  \f}
+  where \f$\mathbf{x}\f$ are the coordinates mapped to by coordinate map.
+  The determinant of the Jacobian matrix is denoted by \f$J\f$.
+- We use \f$s\f$ and \f$t\f$ to be grid point indices. That is, \f$U_s\f$ is
```
How about we use greek indices for these tensor product indices, since latin indices are already overloaded twice (spacetime and spatial) and we have no greek indices around so far? That's what we'll probably do for the elliptic DG paper.

nilsdeppe replied:
You should use whatever is in this document for SXS papers using SpECTRE. That's the point of this document.

nilsvu replied:
Edit: the elliptic dg paper by Trevor, Harald and myself that's in preparation right now

nilsdeppe replied:
Okay, since that's not using SpECTRE I don't think there's an issue with them doing whatever they want :)

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+      \left(dx^i + \beta^i dt\right)\left(dx^j + \beta^j dt\right).
+  \f}
+- The unit timelike normal to the hypersurface is
+  \f$n^a=(1/\alpha,-\beta^i/\alpha)\f$ and \f$n_a=(-\alpha,0,0,0)\f$.
```
I think `t^a` is being used in our documentation right now (e.g. `ComputeGhQuantities`), since `n^i` is already a normal to an element face

nilsdeppe replied:
This was discussed already, we are using this notation.

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+- The fluid pressure, rest mass density, specific internal energy, and specific
+  enthalpy are denoted by \f$p\f$, \f$\rho\f$, \f$\epsilon\f$, and \f$h\f$,
+  respectively.
+- The 4-velocity of the fluid is \f$u^a\f$ and \f$u^a u_a=-1\f$.
```
and -> with

nilsdeppe replied:
I don't think this makes a difference :man_shrugging:

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+  \f}
+- The Faraday tensor is denoted by \f$F^{ab}\f$ and its dual is given by
+  \f$^{*}\!F^{ab}=\frac{1}{2}\epsilon^{abcd}F_{cd}\f$ where
+  \f$\epsilon^{abcd}\f$ is the Levi-Civita tensor.
```
Levi-Civita symbol

nilsdeppe replied:
No, you're wrong.

nilsvu replied:
That‘s a quite unhelpful comment :smile:

If you want to be precise, and don’t want to write _symbol_, then call it a _tensor density_. Only the dedensitized object `\sqrt{g}\epsilon` transforms like a tensor.

nilsdeppe replied:
It's not the symbol, it's the tensor. I don't see how telling you that is unhelpful. The symbol is 0,1,-1, the tensor has a factor of sqrt{-g} in it (see @carmaza's comment). You are asking me to change the equation from something that is correct to something that is incorrect. See, for example, Eq. 5.131 in Baumgarte and Shapiro.

nilsvu replied:
Good, in response to @carmaza's comment you added a definition of the dedensitized tensor.

nilsdeppe replied:
Please don't make it sound like something fundamental changed here. It's the same equation that you thought was incorrect.

nilsvu commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,127 @@
+  \f}
+- For the generalized harmonic evolution system \f$\Pi_{ab}=-n^c\partial_c
+  g_{ab}\f$ and \f$\Phi_{iab}=\partial_i g_{ab}\f$.
+
```
Could you add the energy-momentum tensor and, in particular, the energy density `E=t_a t_b T^{ab}`?

nilsdeppe replied:
Sure, but it's `n_a` and `n_b`.

nilsvu replied:
In https://github.com/sxs-collaboration/spectre/pull/1405#discussion_r264855270 I was told the opposite though. The clash with the face normals is real :smile:

nilsdeppe replied:
Yes, the thing is that SpECTRE is our chance to move to the standard NR notation that is in all the text books. Since the spacetime normal and the element normal show up in completely different contexts I don't think this will be an issue and it is consistent with all the DG literature (maybe except the Jena group who at least in one paper use `s_i` for the element normal). I think it's fine to have this collision. There aren't enough letters to be collision free, unfortunately :P

teukolsky replied:
I'm not thrilled with t^a for the spatial normal to an element interface. First, it has so many other connotations. Second, very soon we will want a notation for a tangential vector to the interface. t^a would be a good choice for that. I'm leaning toward s^a for the normal.

kidder replied:
I don't think anyone is proposing to use t^a as a spatial normal to an element interface.  I thought we had agreed that n_i is the spatial normal to an element interface.

nilsvu commented:
For the record: suggested in another channel to also keep track of _changes_ to these conventions in this document. So if we end up changing a notation, we add a comment to this document that says what changed and when

nilsdeppe committed:
fixup: notation

- Add a table of contents
- Add a history section (currently empty)
- Address comments that I can without getting more feedback.

nilsdeppe commented:
Pushed a fixup addressing the comments that I could and adding a table of contents.

carmaza commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,134 @@
+  \f}
+- The Faraday tensor is denoted by \f$F^{ab}\f$ and its dual is given by
+  \f$^{*}\!F^{ab}=\frac{1}{2}\epsilon^{abcd}F_{cd}\f$ where
+  \f$\epsilon^{abcd}\f$ is the Levi-Civita tensor.
```
...and \epsilon^{0123} = +1/sqrt{-g}, or some other way to introduce the explicit sign convention. (Some GR textbooks define \epsilon_{0123} = +sqrt{-g}, which flips the sign of the fully contravariant \epsilon.)

Might also be better to define this at the beginning, after the metric conventions. Clearly the Levi-Civita tensor isn't restricted to hydro applications.

nilsdeppe replied:
Pushed a fixup. Please let me know if you are happy with the wording :)

carmaza replied:
I'm happy with the change.

teukolsky replied:
I think we should be consistent in following the MTW conventions here, namely \epsilon_{0123} = +sqrt{-g}. The place it makes a difference for us is in the MHD definitions, and there does not seem to be consensus in the literature. Since all our other definitions are those of MTW, let's stick with those.

nilsdeppe committed:
fixup: Add Levi-Civita tensor definition to notation

wthrowe commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,139 @@
+  where \f$N\f$ is the order/degree of the basis in each logical direction.
+- A DG scheme with an \f$N^{\mathrm{th}}\f$ order basis is referred to as a
+  \f$P_N\f$ scheme and converges at order \f$\mathcal{O}\left(\Delta
+  x^{N+1}\right)\f$ where \f$\Delta x\f$ is the 1d size of the element.
```
I think "converges at order O(Dx^{N+1})" is unclear as to whether it is the error term or the last correct term that is proportional to Dx^{N+1}.

wthrowe commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,139 @@
+
+The notation for general relativity quantities used is:
+- The Einstein summation convention is used where repeated indices are summed
+  over.
```
We also use this convention in situations unrelated to GR.  For example, the Rusanov flux written above.

nilsdeppe replied:
Do you think we should have a `General Notation` section at the beginning that this (and anything else general) could go in?

wthrowe replied:
Yes, that sounds good.

wthrowe commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,139 @@
+  \f$^{*}\!F^{ab}=\frac{1}{2}\epsilon^{abcd}F_{cd}\f$.
+- The magnetic field in the frame comoving with the fluid is
+  \f{align*}{
+      b^a = ^{*}\!F^{ab}u_b
```
This is not rendering correctly, although I am not sure why.  You may need a `{}` before the `^`.

wthrowe commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,139 @@
+- The electric and magnetic fields as measured by an Eulerian observer are given
+  by
+  \f{align*}{
+      E^i=&F^{ia}n_a=\alpha F^{0i} \\
```
I think we screw this up all over the place, so I'm not too worried about it, but the align environment is designed to have the alignment marker before the operator (`&=`, not `=&`).  Putting it after results in too little space between the operator and the next symbol.

wthrowe commented on docs/DevGuide/Notation.md:
```diff
@@ -0,0 +1,139 @@
+  \f}
+  while \f$b^2=b^i b_i\f$ is given by
+  \f{align*}{
+      b^2=\frac{B^2}{W}+\left(B^i v_i\right)
```
This doesn't match the paper.

moxcodes commented:
It appears that the ADM notation here differs from a substantial amount of documentation created for the Generalized Harmonic system. Could you make an issue for fixing all of the notation currently in the code base?

nilsdeppe commented:
> It appears that the ADM notation here differs from a substantial amount of documentation created for the Generalized Harmonic system. Could you make an issue for fixing all of the notation currently in the code base?

Added #1514

nilsdeppe committed:
fixup: Address Will's comments

nilsdeppe committed:
fixup: Levi-Civita tensor and symbol sign.

nilsdeppe commented:
@teukolsky would you please check that the sign are what you wanted whenever you get a chance? Thanks!
